src/ILCompiler.ReadyToRun/src/CodeGen/ReadyToRunObjectWriter.cs

```diff
@@ -54,7 +57,21 @@ public void EmitPortableExecutable()
                 var sectionBuilder = new SectionBuilder();
 
                 _headerSectionIndex = sectionBuilder.AddSection(R2RPEBuilder.TextSectionName, SectionCharacteristics.ContainsCode | SectionCharacteristics.MemExecute | SectionCharacteristics.MemRead, 512);
-                sectionBuilder.SetReadyToRunHeaderTable(_nodeFactory.CoreCLRReadyToRunHeader, _nodeFactory.CoreCLRReadyToRunHeader.GetData(_nodeFactory).Data.Length);
+                sectionBuilder.SetReadyToRunHeaderTable(_nodeFactory.Header, _nodeFactory.Header.GetData(_nodeFactory).Data.Length);
+
+                foreach (var depNode in _nodes)
```

Comment:
This node processing loop may become problematic in terms of ensuring dependencies are tracked properly in the graph. By this point in compilation, the graph should be complete and we shouldn't be introducing new dependencies. Once a particular MethodCodeNode is marked by the dependency analysis, it is going into the final image. That's when we should do this ready-to-run table book-keeping.  Perhaps we need a table manager similar to the MetadataManager that hooks into nodes getting marked and configures the appropriate r2r tables. @MichalStrehovsky you implemented the metadata manager - does that sound like a good way to go here?

Reply:
Yes, I was meaning to suggest using the tracking we have in `MetadataManager` for this.

These tables are very much like the "summary nodes" we have when targeting CoreRT. E.g. the ExactMethodInstantiations node: https://github.com/dotnet/corert/blob/5ddf79df02b8be670dc977b5af41b93e65890ad0/src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs#L55

What makes these nodes "summary nodes" is that they don't report any dependencies themselves. They just iterate over things that ended up in the graph as reported by the metadata managed and generate a summary of it.

If they need additional things in the summary (e.g. the exact method instantiation needs some native layout data structures), they have a static method ([example](https://github.com/dotnet/corert/blob/5ddf79df02b8be670dc977b5af41b93e65890ad0/src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs#L108)) that is called from an appropriate place to inject the dependencies on behalf of the table. (This functionality might not be needed at this point.)

Reply:
All of this sounds great to me but I don't think we want the entire R2R implementation to end up in a single PR - could we steer towards finalizing and accepting this delta as an intermediate step so as to unblock additional pending changes, in particular Simon's support for the import sections?

Reply:
I'm fine with you merging this as-is Tomas. I just want to make sure we don't go too deep down this path and agree on a plan.

Reply:
Completely agreed. It seems to me that it might actually make sense to schedule a dedicated technical meeting focused on the next steps to reasonably plan ahead for the next 2-3 weeks and make sure we cover as much stuff as we can before we're asked to cost the remaining work.